tests/raftstore_cases/test_prevote.rs

```diff
@@ -0,0 +1,146 @@
+use std::sync::atomic::AtomicBool;
+use std::sync::mpsc;
+use std::sync::Arc;
+use std::time::Duration;
+
+use super::cluster::{Cluster, Simulator};
+use super::server::new_server_cluster;
+use super::transport_simulate::*;
+use super::util::*;
+use raft::eraftpb::MessageType;
+
+// Validate that prevote is used in elections after a leader is isolated if it is on, or not if it is off.
+fn test_prevote_on_leader_isolation<T: Simulator>(cluster: &mut Cluster<T>, prevote_enabled: bool) {
+    cluster.cfg.raft_store.prevote = prevote_enabled;
+
+    // Setup a notifier
+    let (tx, rx) = mpsc::channel();
+    let response_notifier = Box::new(MessageTypeNotifier::new(
+        MessageType::MsgRequestPreVoteResponse,
+        tx.clone(),
+        Arc::from(AtomicBool::new(true)),
+    ));
+    let request_notifier = Box::new(MessageTypeNotifier::new(
+        MessageType::MsgRequestPreVote,
+        tx.clone(),
+        Arc::from(AtomicBool::new(true)),
+    ));
+
+    // We must start the cluster before adding send filters, otherwise it panics.
+    cluster.run();
+    cluster
+        .sim
+        .write()
+        .unwrap()
+        .add_send_filter(2, response_notifier);
+    cluster
+        .sim
+        .write()
+        .unwrap()
+        .add_send_filter(2, request_notifier);
+
+    // Force an unplanned election.
+    cluster.must_transfer_leader(1, new_peer(1, 1));
+    cluster.add_send_filter(IsolationFilterFactory::new(1));
+
+    // Once we see a response on the wire we know a prevote round is happening.
+    let recieved = rx.recv_timeout(Duration::from_secs(2));
+    assert_eq!(
+        recieved.is_ok(),
+        prevote_enabled,
+        "Recieve a PreVote or PreVoteResponse to a peer when a leader was isolated.",
+    );
+}
+
+#[test]
+fn test_server_prevote_on_leader_isolation() {
+    let mut cluster = new_server_cluster(0, 3);
+    test_prevote_on_leader_isolation(&mut cluster, true);
+}
+
+#[test]
+fn test_server_no_prevote_on_leader_isolation() {
+    let mut cluster = new_server_cluster(0, 3);
+    test_prevote_on_leader_isolation(&mut cluster, false);
+}
+
+fn test_prevote_on_leader_reboot<T: Simulator>(cluster: &mut Cluster<T>, prevote_enabled: bool) {
```

Comment:
What's the difference between it and `test_prevote_on_leader_isolation`?

Reply:
This one tests a node actually stopping and starting again versus just being isolated.